[AutoDiff] Compiler crash when forgetting to import `_Differentiation`

**Describe the bug**
When running the following reproducer, the compiler crashes. It only happens when you forget to import _Differentiation at the top of the file. The compiler emits an error diagnostic before crashing:

**Reproducer**
```swift
// Must be public
public struct Tensor<Scalar> {}

protocol DifferentiableTensorProtocol {
  // Remove `(reverse)` from this modifier on older toolchains
  @differentiable(reverse)
  func annotate()
}

extension Tensor: DifferentiableTensorProtocol
// TensorFlowFloatingPoint is not declared yet
where Scalar: TensorFlowFloatingPoint {
  // Must be public
  public func annotate() {
    fatalError()
  }
}
```
...

**Expected behavior**
Error diagnostic:
```txt
(base) philipturner@M1-Max-MacBook-Pro Desktop % swift file.swift
file.swift:6:4: error: @differentiable attribute used without importing module '_Differentiation'
  @differentiable(reverse)
  ~^~~~~~~~~~~~~~~~~~~~~~~
file.swift:12:15: error: cannot find type 'TensorFlowFloatingPoint' in scope
where Scalar: TensorFlowFloatingPoint {
              ^~~~~~~~~~~~~~~~~~~~~~~
file.swift:10:1: error: type 'Tensor<Scalar>' does not conform to protocol 'DifferentiableTensorProtocol'
extension Tensor: DifferentiableTensorProtocol
^
```

<details>
<summary>Crash</summary>

```
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-5.7-DEVELOPMENT-SNAPSHOT-2022-05-18-a.xctoolchain/usr/bin/swift-frontend -frontend -interpret file.swift -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -color-diagnostics -new-driver-path /Library/Developer/Toolchains/swift-5.7-DEVELOPMENT-SNAPSHOT-2022-05-18-a.xctoolchain/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /Library/Developer/Toolchains/swift-5.7-DEVELOPMENT-SNAPSHOT-2022-05-18-a.xctoolchain/usr/lib/swift -module-name file -target-sdk-version 12.3
1.	Apple Swift version 5.7-dev (LLVM 9fde71b269b42b8, Swift ddc46e1f5bd885d)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "file.swift")
4.	While type-checking extension of Tensor (at file.swift:10:1)
5.	While type-checking protocol conformance to 'DifferentiableTensorProtocol' (at file.swift:4:1) for type 'Tensor<Scalar>' (declared at [file.swift:2:8 - line:2:31] RangeText="struct Tensor<Scalar> {")
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000106ada01c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x0000000106ad927c llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x0000000106ada680 SignalHandler(int) + 304
3  libsystem_platform.dylib 0x000000019923f4a4 _sigtramp + 56
4  swift-frontend           0x00000001032907a8 swift::TypeChecker::conformsToProtocol(swift::Type, swift::ProtocolDecl*, swift::ModuleDecl*, bool) + 52
5  swift-frontend           0x00000001032907a8 swift::TypeChecker::conformsToProtocol(swift::Type, swift::ProtocolDecl*, swift::ModuleDecl*, bool) + 52
6  swift-frontend           0x00000001031db748 swift::TypeChecker::inferDifferentiabilityParameters(swift::AbstractFunctionDecl*, swift::GenericEnvironment*) + 896
7  swift-frontend           0x000000010329d778 diagnoseMatch(swift::ModuleDecl*, swift::NormalProtocolConformance*, swift::ValueDecl*, swift::RequirementMatch const&) + 2288
8  swift-frontend           0x00000001032a4d28 std::__1::__function::__func<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_11, std::__1::allocator<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_11>, void (swift::NormalProtocolConformance*)>::operator()(swift::NormalProtocolConformance*&&) + 92
9  swift-frontend           0x000000010328dc94 swift::ConformanceChecker::diagnoseOrDefer(swift::ValueDecl const*, bool, std::__1::function<void (swift::NormalProtocolConformance*)>) + 216
10 swift-frontend           0x0000000103289b38 swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind) + 1384
11 swift-frontend           0x000000010328b5f0 swift::ConformanceChecker::checkConformance(swift::MissingWitnessDiagnosisKind) + 1216
12 swift-frontend           0x00000001032889a4 swift::MultiConformanceChecker::checkIndividualConformance(swift::NormalProtocolConformance*, bool) + 3440
13 swift-frontend           0x0000000103287a34 swift::MultiConformanceChecker::checkAllConformances() + 120
14 swift-frontend           0x0000000103295168 swift::TypeChecker::checkConformancesInContext(swift::IterableDeclContext*) + 4948
15 swift-frontend           0x0000000103251d3c swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) + 8588
16 swift-frontend           0x000000010324c6e0 (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 224
17 swift-frontend           0x000000010324c5ec swift::TypeChecker::typeCheckDecl(swift::Decl*, bool) + 116
18 swift-frontend           0x00000001032fcee4 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 176
19 swift-frontend           0x00000001032fed7c llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 400
20 swift-frontend           0x00000001032feb1c llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 124
21 swift-frontend           0x00000001032fcd28 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 44
22 swift-frontend           0x000000010265a9a8 bool llvm::function_ref<bool (swift::SourceFile&)>::callback_fn<swift::CompilerInstance::performSema()::$_7>(long, swift::SourceFile&) + 16
23 swift-frontend           0x0000000102656e64 swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<bool (swift::SourceFile&)>) + 160
24 swift-frontend           0x0000000102656d98 swift::CompilerInstance::performSema() + 76
25 swift-frontend           0x00000001026087a4 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) + 60
26 swift-frontend           0x00000001025fd108 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3388
27 swift-frontend           0x0000000102511bc0 swift::mainEntry(int, char const**) + 3764
28 dyld                     0x000000010f51908c start + 520
zsh: segmentation fault  swift file.swift
```

</details>

**Environment (please complete the following information):**
 - OS: macOS 12.4
 - Toolchain: May 18, 2022 Swift 5.7 Development Snapshot. This diverges from the main branch and nightly builds as we prepare to release Swift 5.7, but it still includes the _Differentiation module (unlike true release toolchains).

This crash has existed on older toolchains, as far back as January 2020. The stack traces change significantly over time.

<details>
<summary>January 29, 2020</summary>

```
(base) philipturner@M1-Max-MacBook-Pro Desktop % swift file.swift
file.swift:6:4: error: differentiable programming is an experimental feature that is currently disabled
  @differentiable
  ~^~~~~~~~~~~~~~
file.swift:12:15: error: use of undeclared type 'TensorFlowFloatingPoint'
where Scalar: TensorFlowFloatingPoint {
              ^~~~~~~~~~~~~~~~~~~~~~~
file.swift:10:1: error: type 'Tensor<Scalar>' does not conform to protocol 'DifferentiableTensorProtocol'
extension Tensor: DifferentiableTensorProtocol
^
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2020-01-29-a.xctoolchain/usr/bin/swift -frontend -interpret file.swift -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -color-diagnostics -module-name file 
1.	Apple Swift version 5.2-dev
2.	While evaluating request TypeCheckSourceFileRequest(source_file "file.swift", 0)
3.	While type-checking extension of Tensor (at file.swift:10:1)
4.	While type-checking protocol conformance to 'DifferentiableTensorProtocol' (at file.swift:4:1) for type 'Tensor<Scalar>' (declared at [file.swift:2:8 - line:2:31] RangeText="struct Tensor<Scalar> {")
0  swift                    0x0000000108502425 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x0000000108501635 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x0000000108502a16 SignalHandler(int) + 278
3  libsystem_platform.dylib 0x00007ff81a3aadfd _sigtramp + 29
4  libsystem_platform.dylib 000000000000000000 _sigtramp + 3854914080
5  swift                    0x0000000105304337 conformsToDifferentiable(swift::Type, swift::DeclContext*) + 87
6  swift                    0x00000001056b3140 llvm::Expected<swift::DifferentiableAttributeTypeCheckRequest::OutputType> swift::Evaluator::getResultCached<swift::DifferentiableAttributeTypeCheckRequest, (void*)0>(swift::DifferentiableAttributeTypeCheckRequest const&) + 32
7  swift                    0x00000001056b1b13 swift::DifferentiableAttributeTypeCheckRequest::OutputType swift::evaluateOrDefault<swift::DifferentiableAttributeTypeCheckRequest>(swift::Evaluator&, swift::DifferentiableAttributeTypeCheckRequest, swift::DifferentiableAttributeTypeCheckRequest::OutputType) + 35
8  swift                    0x000000010539912c diagnoseMatch(swift::ModuleDecl*, swift::NormalProtocolConformance*, swift::ValueDecl*, swift::RequirementMatch const&) + 4380
9  swift                    0x00000001053a0f3a std::__1::__function::__func<swift::ConformanceChecker::resolveWitnessViaLookup(swift::ValueDecl*)::$_21, std::__1::allocator<swift::ConformanceChecker::resolveWitnessViaLookup(swift::ValueDecl*)::$_21>, void (swift::NormalProtocolConformance*)>::operator()(swift::NormalProtocolConformance*&&) + 106
10 swift                    0x000000010538b35a swift::ConformanceChecker::diagnoseOrDefer(swift::ValueDecl*, bool, std::__1::function<void (swift::NormalProtocolConformance*)>) + 202
11 swift                    0x000000010538d803 swift::ConformanceChecker::resolveWitnessViaLookup(swift::ValueDecl*) + 3107
12 swift                    0x000000010538ece1 swift::ConformanceChecker::resolveValueWitnesses() + 385
13 swift                    0x0000000105389979 swift::ConformanceChecker::checkConformance(swift::MissingWitnessDiagnosisKind) + 313
14 swift                    0x0000000105388587 swift::MultiConformanceChecker::checkIndividualConformance(swift::NormalProtocolConformance*, bool) + 8839
15 swift                    0x0000000105386192 swift::MultiConformanceChecker::checkAllConformances() + 114
16 swift                    0x000000010539164d swift::TypeChecker::checkConformancesInContext(swift::DeclContext*, swift::IterableDeclContext*) + 5165
17 swift                    0x000000010535b631 (anonymous namespace)::DeclChecker::visitExtensionDecl(swift::ExtensionDecl*) + 2881
18 swift                    0x0000000105354dc3 (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 1747
19 swift                    0x00000001053546df swift::TypeChecker::typeCheckDecl(swift::Decl*) + 63
20 swift                    0x00000001053ebd08 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*, unsigned int) const + 408
21 swift                    0x00000001053bb675 swift::SimpleRequest<swift::TypeCheckSourceFileRequest, bool (swift::SourceFile*, unsigned int), (swift::CacheKind)2>::evaluateRequest(swift::TypeCheckSourceFileRequest const&, swift::Evaluator&) + 21
22 swift                    0x00000001053edde4 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 356
23 swift                    0x00000001053edba8 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 56
24 swift                    0x00000001053eba65 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 37
25 swift                    0x0000000104aecf05 swift::CompilerInstance::parseAndTypeCheckMainFileUpTo(swift::SourceFile::ASTStage_t) + 533
26 swift                    0x0000000104aebb72 swift::CompilerInstance::parseAndCheckTypesUpTo(swift::CompilerInstance::ImplicitImports const&, swift::SourceFile::ASTStage_t) + 418
27 swift                    0x0000000104aeb351 swift::CompilerInstance::performSemaUpTo(swift::SourceFile::ASTStage_t) + 625
28 swift                    0x00000001048a7e11 performCompile(swift::CompilerInstance&, swift::CompilerInvocation const&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 1905
29 swift                    0x00000001048a6713 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 5139
30 swift                    0x000000010483b88e main + 734
31 dyld                     0x000000020ede251e start + 462
32 dyld                     0x0000000000000003 start + 4045528243
zsh: segmentation fault  swift file.swift
```

</details>



<details>
<summary>July 22, 2020</summary>

```
(base) philipturner@M1-Max-MacBook-Pro Desktop % swift file.swift
file.swift:6:4: error: @differentiable attribute used without importing module '_Differentiation'
  @differentiable
  ~^~~~~~~~~~~~~~
file.swift:12:15: error: cannot find type 'TensorFlowFloatingPoint' in scope
where Scalar: TensorFlowFloatingPoint {
              ^~~~~~~~~~~~~~~~~~~~~~~
file.swift:10:1: error: type 'Tensor<Scalar>' does not conform to protocol 'DifferentiableTensorProtocol'
extension Tensor: DifferentiableTensorProtocol
^
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2020-07-22-a.xctoolchain/usr/bin/swift-frontend -frontend -interpret file.swift -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -color-diagnostics -target-sdk-version 12.3 -module-name file 
1.	Apple Swift version 5.3-dev (LLVM e966de21c9ff9ed, Swift 493b4a8ef836af0)
2.	While evaluating request TypeCheckSourceFileRequest(source_file "file.swift")
3.	While type-checking extension of Tensor (at file.swift:10:1)
4.	While type-checking protocol conformance to 'DifferentiableTensorProtocol' (at file.swift:4:1) for type 'Tensor<Scalar>' (declared at [file.swift:2:8 - line:2:31] RangeText="struct Tensor<Scalar> {")
0  swift-frontend           0x000000010470f345 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift-frontend           0x000000010470e5c5 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x000000010470f916 SignalHandler(int) + 262
3  libsystem_platform.dylib 0x00007ff81a3aadfd _sigtramp + 29
4  libsystem_platform.dylib 000000000000000000 _sigtramp + 18446603370140881440
5  libsystem_platform.dylib 0x000000031026aab8 _sigtramp + 18446603383296752856
zsh: segmentation fault  swift file.swift
```

</details>


<details>
<summary>January 27, 2021</summary>

```
(base) philipturner@M1-Max-MacBook-Pro Desktop % swift file.swift
file.swift:6:4: error: @differentiable attribute used without importing module '_Differentiation'
  @differentiable
  ~^~~~~~~~~~~~~~
file.swift:12:15: error: cannot find type 'TensorFlowFloatingPoint' in scope
where Scalar: TensorFlowFloatingPoint {
              ^~~~~~~~~~~~~~~~~~~~~~~
file.swift:10:1: error: type 'Tensor<Scalar>' does not conform to protocol 'DifferentiableTensorProtocol'
extension Tensor: DifferentiableTensorProtocol
^
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-01-27-a.xctoolchain/usr/bin/swift-frontend -frontend -interpret file.swift -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -color-diagnostics -target-sdk-version 12.3 -module-name file 
1.	Apple Swift version 5.4-dev (LLVM e2976fe639d1f50, Swift f9be289b9e04b3d)
2.	While evaluating request TypeCheckSourceFileRequest(source_file "file.swift")
3.	While type-checking extension of Tensor (at file.swift:10:1)
4.	While type-checking protocol conformance to 'DifferentiableTensorProtocol' (at file.swift:4:1) for type 'Tensor<Scalar>' (declared at [file.swift:2:8 - line:2:31] RangeText="struct Tensor<Scalar> {")
0  swift-frontend           0x000000010662f9fc llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 52
1  swift-frontend           0x000000010662ec5c llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x000000010662ffe4 SignalHandler(int) + 292
3  libsystem_platform.dylib 0x000000019923f4a4 _sigtramp + 56
4  swift-frontend           0x0000000103599248 diagnoseMatch(swift::ModuleDecl*, swift::NormalProtocolConformance*, swift::ValueDecl*, swift::RequirementMatch const&) + 2668
5  swift-frontend           0x0000000103599248 diagnoseMatch(swift::ModuleDecl*, swift::NormalProtocolConformance*, swift::ValueDecl*, swift::RequirementMatch const&) + 2668
6  swift-frontend           0x000000010359f898 std::__1::__function::__func<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_9, std::__1::allocator<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_9>, void (swift::NormalProtocolConformance*)>::operator()(swift::NormalProtocolConformance*&&) + 92
7  swift-frontend           0x000000010358a9c0 swift::ConformanceChecker::diagnoseOrDefer(swift::ValueDecl*, bool, std::__1::function<void (swift::NormalProtocolConformance*)>) + 208
8  swift-frontend           0x0000000103587468 swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind) + 1568
9  swift-frontend           0x0000000103588c54 swift::ConformanceChecker::checkConformance(swift::MissingWitnessDiagnosisKind) + 952
10 swift-frontend           0x0000000103586c4c swift::MultiConformanceChecker::checkIndividualConformance(swift::NormalProtocolConformance*, bool) + 5632
11 swift-frontend           0x0000000103585444 swift::MultiConformanceChecker::checkAllConformances() + 120
12 swift-frontend           0x0000000103590d4c swift::TypeChecker::checkConformancesInContext(swift::IterableDeclContext*) + 3972
13 swift-frontend           0x00000001035535a8 swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) + 7172
14 swift-frontend           0x0000000103551404 (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 168
15 swift-frontend           0x0000000103551348 swift::TypeChecker::typeCheckDecl(swift::Decl*) + 132
16 swift-frontend           0x00000001035eec30 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 164
17 swift-frontend           0x00000001035efc80 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 396
18 swift-frontend           0x00000001035efa24 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 120
19 swift-frontend           0x00000001035eea94 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 44
20 swift-frontend           0x0000000102a6bc8c swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<void (swift::SourceFile&)>) + 144
21 swift-frontend           0x0000000102a6bbd0 swift::CompilerInstance::performSema() + 76
22 swift-frontend           0x0000000102977384 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>) + 56
23 swift-frontend           0x000000010296f1a0 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4292
24 swift-frontend           0x000000010291cc88 main + 784
25 dyld                     0x000000010dcb108c start + 520
zsh: segmentation fault  swift file.swift
```

</details>


<details>
<summary>April 18, 2021</summary>

```
(base) philipturner@M1-Max-MacBook-Pro Desktop % swift file.swift
file.swift:6:4: error: @differentiable attribute used without importing module '_Differentiation'
  @differentiable(reverse)
  ~^~~~~~~~~~~~~~~~~~~~~~~
file.swift:12:15: error: cannot find type 'TensorFlowFloatingPoint' in scope
where Scalar: TensorFlowFloatingPoint {
              ^~~~~~~~~~~~~~~~~~~~~~~
file.swift:10:1: error: type 'Tensor<Scalar>' does not conform to protocol 'DifferentiableTensorProtocol'
extension Tensor: DifferentiableTensorProtocol
^
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-04-18-a.xctoolchain/usr/bin/swift-frontend -frontend -interpret file.swift -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -color-diagnostics -new-driver-path /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-04-18-a.xctoolchain/usr/bin/swift-driver -resource-dir /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-04-18-a.xctoolchain/usr/lib/swift -module-name file -target-sdk-version 12.3.0
1.	Apple Swift version 5.5-dev (LLVM fb5a91b4e892ed9, Swift f17142883421ec0)
2.	While evaluating request TypeCheckSourceFileRequest(source_file "file.swift")
3.	While type-checking extension of Tensor (at file.swift:10:1)
4.	While type-checking protocol conformance to 'DifferentiableTensorProtocol' (at file.swift:4:1) for type 'Tensor<Scalar>' (declared at [file.swift:2:8 - line:2:31] RangeText="struct Tensor<Scalar> {")
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000108ea7e87 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x0000000108ea7045 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x0000000108ea84a6 SignalHandler(int) + 262
3  libsystem_platform.dylib 0x00007ff81a3aadfd _sigtramp + 29
4  libsystem_platform.dylib 0x0000000319515f90 _sigtramp + 18446603383450546608
5  swift-frontend           0x000000010527224b swift::TypeChecker::inferDifferentiabilityParameters(swift::AbstractFunctionDecl*, swift::GenericEnvironment*) + 1179
6  swift-frontend           0x0000000105333ed3 diagnoseMatch(swift::ModuleDecl*, swift::NormalProtocolConformance*, swift::ValueDecl*, swift::RequirementMatch const&) + 1267
7  swift-frontend           0x000000010533c6ea std::__1::__function::__func<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_10, std::__1::allocator<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_10>, void (swift::NormalProtocolConformance*)>::operator()(swift::NormalProtocolConformance*&&) + 90
8  swift-frontend           0x0000000105323527 swift::ConformanceChecker::diagnoseOrDefer(swift::ValueDecl*, bool, std::__1::function<void (swift::NormalProtocolConformance*)>) + 231
9  swift-frontend           0x000000010531f58a swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind) + 1914
10 swift-frontend           0x0000000105321249 swift::ConformanceChecker::checkConformance(swift::MissingWitnessDiagnosisKind) + 1177
11 swift-frontend           0x000000010531de14 swift::MultiConformanceChecker::checkIndividualConformance(swift::NormalProtocolConformance*, bool) + 3716
12 swift-frontend           0x000000010531cce9 swift::MultiConformanceChecker::checkAllConformances() + 169
13 swift-frontend           0x000000010532aaa0 swift::TypeChecker::checkConformancesInContext(swift::IterableDeclContext*) + 4960
14 swift-frontend           0x00000001052dc9e4 swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) + 9140
15 swift-frontend           0x00000001052d86b4 (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 228
16 swift-frontend           0x00000001052d85b2 swift::TypeChecker::typeCheckDecl(swift::Decl*) + 130
17 swift-frontend           0x0000000105396818 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 184
18 swift-frontend           0x000000010539789d llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 429
19 swift-frontend           0x0000000105397622 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 114
20 swift-frontend           0x000000010539665e swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 46
21 swift-frontend           0x00000001045f018e swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<void (swift::SourceFile&)>) + 174
22 swift-frontend           0x00000001045f002a swift::CompilerInstance::performSema() + 74
23 swift-frontend           0x00000001044c34ad swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4877
24 swift-frontend           0x0000000104413386 main + 566
25 dyld                     0x000000021105a51e start + 462
26 dyld                     0x0000000211055000 start + 18446744073709530288
zsh: segmentation fault  swift file.swift
```

</details>



<details>
<summary>July 7, 2021</summary>

```
(base) philipturner@M1-Max-MacBook-Pro Desktop % swift file.swift
file.swift:6:4: error: @differentiable attribute used without importing module '_Differentiation'
  @differentiable(reverse)
  ~^~~~~~~~~~~~~~~~~~~~~~~
file.swift:12:15: error: cannot find type 'TensorFlowFloatingPoint' in scope
where Scalar: TensorFlowFloatingPoint {
              ^~~~~~~~~~~~~~~~~~~~~~~
file.swift:10:1: error: type 'Tensor<Scalar>' does not conform to protocol 'DifferentiableTensorProtocol'
extension Tensor: DifferentiableTensorProtocol
^
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-07-07-a.xctoolchain/usr/bin/swift-frontend -frontend -interpret file.swift -clang-target arm64-apple-macosx12.0 -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -color-diagnostics -new-driver-path /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-07-07-a.xctoolchain/usr/bin/swift-driver -resource-dir /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-07-07-a.xctoolchain/usr/lib/swift -module-name file -target-sdk-version 12.3.0
1.	Apple Swift version 5.5-dev (LLVM 341fda9ea021b38, Swift 753474e206d06d1)
2.	
3.	While evaluating request TypeCheckSourceFileRequest(source_file "file.swift")
4.	While type-checking extension of Tensor (at file.swift:10:1)
5.	While type-checking protocol conformance to 'DifferentiableTensorProtocol' (at file.swift:4:1) for type 'Tensor<Scalar>' (declared at [file.swift:2:8 - line:2:31] RangeText="struct Tensor<Scalar> {")
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000108b04914 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x0000000108b03b74 llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x0000000108b04f6c SignalHandler(int) + 292
3  libsystem_platform.dylib 0x000000019923f4a4 _sigtramp + 56
4  swift-frontend           0x00000001057f2434 swift::TypeChecker::conformsToProtocol(swift::Type, swift::ProtocolDecl*, swift::ModuleDecl*) + 36
5  swift-frontend           0x00000001057f2434 swift::TypeChecker::conformsToProtocol(swift::Type, swift::ProtocolDecl*, swift::ModuleDecl*) + 36
6  swift-frontend           0x000000010574d780 swift::TypeChecker::inferDifferentiabilityParameters(swift::AbstractFunctionDecl*, swift::GenericEnvironment*) + 1080
7  swift-frontend           0x00000001057fe3fc diagnoseMatch(swift::ModuleDecl*, swift::NormalProtocolConformance*, swift::ValueDecl*, swift::RequirementMatch const&) + 1620
8  swift-frontend           0x0000000105805710 std::__1::__function::__func<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_10, std::__1::allocator<swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind)::$_10>, void (swift::NormalProtocolConformance*)>::operator()(swift::NormalProtocolConformance*&&) + 92
9  swift-frontend           0x00000001057ef880 swift::ConformanceChecker::diagnoseOrDefer(swift::ValueDecl*, bool, std::__1::function<void (swift::NormalProtocolConformance*)>) + 216
10 swift-frontend           0x00000001057ec0d4 swift::ConformanceChecker::diagnoseMissingWitnesses(swift::MissingWitnessDiagnosisKind) + 1336
11 swift-frontend           0x00000001057ed8f0 swift::ConformanceChecker::checkConformance(swift::MissingWitnessDiagnosisKind) + 984
12 swift-frontend           0x00000001057eaf78 swift::MultiConformanceChecker::checkIndividualConformance(swift::NormalProtocolConformance*, bool) + 3256
13 swift-frontend           0x00000001057ea08c swift::MultiConformanceChecker::checkAllConformances() + 120
14 swift-frontend           0x00000001057f6194 swift::TypeChecker::checkConformancesInContext(swift::IterableDeclContext*) + 4644
15 swift-frontend           0x00000001057af0d8 swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) + 7528
16 swift-frontend           0x00000001057abde8 (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 224
17 swift-frontend           0x00000001057abcf8 swift::TypeChecker::typeCheckDecl(swift::Decl*) + 104
18 swift-frontend           0x0000000105855440 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 160
19 swift-frontend           0x0000000105856570 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 396
20 swift-frontend           0x0000000105856314 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 124
21 swift-frontend           0x0000000105855294 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 44
22 swift-frontend           0x0000000104c9a190 swift::CompilerInstance::forEachFileToTypeCheck(llvm::function_ref<void (swift::SourceFile&)>) + 144
23 swift-frontend           0x0000000104c9a0d4 swift::CompilerInstance::performSema() + 76
24 swift-frontend           0x0000000104b95670 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>) + 56
25 swift-frontend           0x0000000104b8cd08 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4148
26 swift-frontend           0x0000000104af8fcc main + 500
27 dyld                     0x0000000110a4508c start + 520
zsh: segmentation fault  swift file.swift
```

</details>

**Other commentary**

I found this while trying to build S4TF on a release toolchain. The error is resolved after adding [philipturner/differentiation](https://github.com/philipturner/differentiation) as a Swift package dependency.

--- a/lib/Sema/TypeCheckAttr.cpp
+++ b/lib/Sema/TypeCheckAttr.cpp
@@ -5780,8 +5780,10 @@ IndexSubset *DifferentiableAttributeTypeCheckRequest::evaluate(
 
   // `@differentiable` attribute requires experimental differentiable
   // programming to be enabled.
-  if (checkIfDifferentiableProgrammingEnabled(attr, D))
+  if (checkIfDifferentiableProgrammingEnabled(attr, D)) {
+    attr->setInvalid();
     return nullptr;
+  }
 
   // If `@differentiable` attribute is declared directly on a
   // `AbstractStorageDecl` (a stored/computed property or subscript),
@@ -5791,8 +5793,10 @@ IndexSubset *DifferentiableAttributeTypeCheckRequest::evaluate(
 
   // Resolve the original `AbstractFunctionDecl`.
   auto *original = resolveDifferentiableAttrOriginalFunction(attr);
-  if (!original)
+  if (!original) {
+    attr->setInvalid();
     return nullptr;
+  }
 
   return typecheckDifferentiableAttrforDecl(original, attr);
 }

--- a/test/AutoDiff/Sema/differentiable_swift_without_import.swift
+++ b/test/AutoDiff/Sema/differentiable_swift_without_import.swift
@@ -1,0 +1,16 @@
+// RUN: %target-typecheck-verify-swift
+
+protocol P {
+  @differentiable(reverse) // expected-error {{@differentiable attribute used without importing module '_Differentiation'}}
+  func req()
+}
+
+struct S: P {
+  // Had the `_Differentiation` module been imported, this conformance 
+  // requirement would have generated an error, complaining about the 
+  // missing `@derivative` attribute. However, because the `_Differentiation`
+  // module import is missing, the `@differential` attribute on `P` is marked
+  // invalid and therefore not used while validating witnesses (specifically `S.req()`) 
+  // for the `P.req()` requirement.
+  public func req() {}
+}

